framework/include/utils/MooseRandomStateless.h

```diff
@@ -0,0 +1,386 @@
+    {
+    }
+
+    /// @return Reference to the uniform Real generator
+    const Generator<Real> & getRandGenerator() const { return _rand_generator; }
+
+    /// @return Reference to the full-range integer generator
+    const Generator<unsigned int> & getRandlGenerator() const { return _randl_generator; }
+
+    /**
+     * @brief Get or create a bounded integer generator for a given range.
+     *
+     * Creates a new generator if one for this range does not exist.
+     *
+     * @param range The integer range of bounded integer (upper - lower)
+     * @return Reference to the corresponding Generator instance
+     */
+    const Generator<unsigned int> & getRandlbGenerator(unsigned int range) const
+    {
+      auto rng_func = [&range](mt_state * state) { return rds_iuniform(state, 0, range); };
+      auto [it, is_new] = _randlb_generators.try_emplace(range, rng_func, _seed);
+      if (is_new)
+        it->second.advance(_advance_count);
+      return it->second;
+    }
+
+    /**
+     * @brief Advance the RNGs by a fixed count.
+     *
+     * @param count Number of RNG calls to skip
+     */
+    void advance(std::size_t count)
+    {
+      _rand_generator.advance(count);
+      _randl_generator.advance(count);
+      for (auto & [range, gen] : _randlb_generators)
+        gen.advance(count);
+      _advance_count += count;
+    }
+
+  private:
+    /// Seed shared by all internal generators
+    const unsigned int _seed;
+    /// Uniform Real [0, 1)
+    Generator<Real> _rand_generator;
+    /// Uniform uint32
+    Generator<unsigned int> _randl_generator;
+    /// Bounded uniform uint32 (indexed based on bounding range)
+    mutable std::map<unsigned int, Generator<unsigned int>> _randlb_generators;
+
+    /// The number of counts the generators have advanced
+    /// This needs to be kept around for generation of new randlb generators
+    std::size_t _advance_count = 0;
+
+    // for restart capability
+    friend void dataStore<MooseRandomStateless>(std::ostream & stream,
+                                                MooseRandomStateless & v,
+                                                void * context);
+    friend void
+    dataLoad<MooseRandomStateless>(std::istream & stream, MooseRandomStateless & v, void * context);
+  };
+
+  /**
+   * @brief Retrieve an existing generator for a given stream index.
+   *
+   * @param i Stream index
+   * @return Reference to the associated `GeneratorHandler`
+   * @throws MooseError if no generator has been seeded for this index
+   */
+  const GeneratorHandler & getGenerator(std::size_t i) const
+  {
+    auto it = _generators.find(i);
+    if (it == _generators.end())
+      mooseError("No random generator initialized for id: ", i);
+    return *it->second;
+  }
+
+  /// Map of per-stream generators (indexed by stream ID)
+  std::map<std::size_t, std::unique_ptr<GeneratorHandler>> _generators;
```

Comment:
```suggestion
  std::unordered_map<std::size_t, std::unique_ptr<GeneratorHandler>> _generators;
```

Hashing for std::size_t is hella fast

Reply:
I figured the critical operation is the map lookup, not the insertion; which is O(n) for unordered_map and O(log n) for map. So wouldn't changing to unordered_map be slower for the lookup?

Reply:
lookup should be O(constant) for unordered_map ?